Allow single dimension setter for mb.Box class

daico007 opened the pull request:
### PR Summary:
Patches for #365. Makes `box._lengths`, `box._maxs` and `box._mins` to be "immutable". You still can change mins, maxs, and lengths by `box.mins = [1, 1, 1]` (which go through the `settter` method), but `box.mins[0] = 1` will return a `ValueError`.
 
### PR Checklist
------------
 - [ ] Includes appropriate unit test(s)
 - [ ] Appropriate docstring(s) are added/updated
 - [ ] Code is (approximately) PEP8 compliant
 - [ ] Issue(s) raised/addressed?

daico007 commented:
So the problem now is that if you try to do the illegal thing like `box_mins[0] = 1`, the error we will receive is `ValueError: assignment destination is read-only`. So the next thing to do is trying to give a more helpful error message (like a direction to properly change mins, maxs, and lengths). The method I am thinking of is trying to catch when python tries to give a `ValueError` and use the try-except thingy to give our own error message. However, I am not sure how to catch that `ValueError`.
 
It seems like when we do `box.mins[0] = 1`, it will call the getter method, and the error will happen when python tries to do the assignment next (this part I am not really sure about). So is there a way we can target that step and alter the error message. 

(@umesh-timalsina pretty sure I gonna need your expertise here)

umesh-timalsina commented:
@daico007 I am on it.

umesh-timalsina commented:
@daico007 Are trying to catch the `ValueError` that comes from numpy?

daico007 commented:
> @daico007 Are trying to catch the `ValueError` that comes from numpy?

Yes

daico007 commented:
Some tests are failing now, so I am trying to figure that out. But yeah, I am trying to catch that numpy `ValueError`.

umesh-timalsina commented:
>Some tests are failing now, so I am trying to figure that out. But yeah, I am trying to catch that numpy ValueError.

That won't be so straight forward because nowhere in the codebase is the `ValueError` raised. I would do away by adding a `Raises` in the `docstring` for the class. Basically, to catch an exception, it must be raised somewhere in the code. But, since all our code does is exposes the numpy array object (with writable flag turned off) there might be no control on it at the package level.

mattwthompson commented:
I'm being overly picky here but the setters should also be able to take in tuples (I think you only need to modify the `isinstance` lines) and be tested like so

```python3
>>> import mbuild as mb
>>> box = mb.Box([4, 4, 4])
>>> box.maxs = (5, 5, 5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mwt/software/mbuild/mbuild/box.py", line 74, in maxs
    assert maxes.shape == (3, )
AttributeError: 'tuple' object has no attribute 'shape'
>>>

daico007 commented:
I removed the `isinstance` part, so now all the setters should also accept tuple.
At this point, users won't be able to edit a single dimension of either the box min, max or length (since they are all immutable). 
However, there is still one issue that need to be solved: users can still do this `box.mins[0] = 4`. Although that will not change the `box._mins`, however, there is no error raised either. So I am still thinking of a way to insert a warning or error when users trying to do those illegal changes.

mattwthompson commented:
That's what I'm getting, too

```python3
>>> import mbuild as mb
>>> box = mb.Box([4, 4, 4])
>>> box
Box(mins=[0. 0. 0.], maxs=[4. 4. 4.], angles=[90. 90. 90.])
>>> box.lengths = (5, 5, 5)
>>> box
Box(mins=[-0.5 -0.5 -0.5], maxs=[4.5 4.5 4.5], angles=[90. 90. 90.])
>>> box.maxs[0] = 40
>>> box
Box(mins=[-0.5 -0.5 -0.5], maxs=[4.5 4.5 4.5], angles=[90. 90. 90.])
```

Of the options available, I agree that disallowing setting a single element is the best option. The worst case is needing to do `box.maxs = [40, box.maxs[1], boxs.maxs[2]]` . Although I'm not sure how to properly raise an exception since the setter doesn't get called on `box.lengths[0]`, i.e. this change won't even raise an exception
```
     @lengths.setter
     def lengths(self, lengths):
         raise ValueError('foobar')
         lengths = np.array(lengths, dtype=np.float)
         assert lengths.shape == (3, )

daico007 commented:
So subclass the `np.ndarray` as `boxArray` with a few more attributes. Now the users could change just one dimension of the `box.mins`, `box.maxs`,`box.lengths`. 
This is currently a prototype, so it may be a bit inefficient. Please let me know what do you guys think, and if there is anything I can change to make this more efficient.

mattwthompson commented on mbuild/tests/test_box.py:
```diff
@@ -72,3 +72,18 @@ def test_setters_with_lists(self):
+        box.lengths[1] = 6
+        assert (box.mins == np.array([0, -1, 1], dtype=np.float)).all()
+        assert (box.maxs == np.array([5, 5, 4], dtype=np.float)).all()
+        assert (box.lengths == np.array([5, 6, 3], dtype=np.float)).all()
```
These are probably fine but I think `np.allclose` is safer?

mattwthompson commented on mbuild/box.py:
```diff
@@ -61,35 +61,53 @@ def angles(self):
     def __repr__(self):
         return "Box(mins={}, maxs={}, angles={})".format(self.mins, self.maxs, self.angles)
+
+class boxArray(np.ndarray):
```
Extremely nitpicky but PEP8 recommends CapCase for class names

https://www.python.org/dev/peps/pep-0008/#class-names

mattwthompson commented on mbuild/box.py:
```diff
@@ -24,27 +24,27 @@ def __init__(self, lengths=None, mins=None, maxs=None, angles=None):
 
     @property
-    def mins(self):
+    def mins(self): 
```
Extra space

daico007 commented:
@uppittu11  Good catch! I just added an `assert` check for `mins` and `maxs` in the `init` step. I also added a few lines so that when only one value is given for `lengths`, we will by default create a cube of the given length.
``` 
>>> from mbuild import Box
>>> box = Box(4)
>>>box
Box(mins=[0. 0. 0.], maxs=[4. 4. 4.], angles=[90. 90. 90.])
```

uppittu11 commented on mbuild/box.py:
```diff
@@ -24,23 +24,32 @@ def __init__(self, lengths=None, mins=None, maxs=None, angles=None):
+            angles = BoxArray(array=(90.0, 90.0, 90.0), var="angles", box=self)
         elif isinstance(angles, (list, np.ndarray)):
-            angles = np.array(angles, dtype=np.float)
+            angles = BoxArray(array=angles, var="anglese", box=self)
```
There's a typo.
```suggestion
            angles = BoxArray(array=angles, var="angles", box=self)
```

uppittu11 commented on mbuild/box.py:
```diff
@@ -24,23 +24,32 @@ def __init__(self, lengths=None, mins=None, maxs=None, angles=None):
+            assert lengths.shape == (3, )
+            self._mins = BoxArray(array=(0,0,0), var="mins", box=self)
+            self._maxs = BoxArray(array=lengths, var="maxs", box=self)
+            self._lengths = BoxArray(array=lengths, var="lenghts", box=self)
```
Typo
```suggestion
            self._lengths = BoxArray(array=lengths, var="lengths", box=self)
```

uppittu11 commented on mbuild/box.py:
```diff
@@ -2,7 +2,7 @@
 import numpy as np
 
-__all__ = ['_BoxArray']
+__all__ = ['Box', '_BoxArray']
```
Kind of a nit, but `_BoxArray` shouldn't be included in `__all__` since it isn't supposed to be used outside of this module. 

For example, this shouldn't be allowed:
```python3
>>> from mbuild.box import *
>>> _BoxArray
<class 'mbuild.box._BoxArray'>
```

uppittu11 commented on mbuild/formats/lammpsdata.py:
```diff
@@ -142,7 +142,6 @@ def write_lammpsdata(structure, filename, atom_style='full',
     box = Box(lengths=np.array([0.1 * val for val in structure.box[0:3]]),
               angles=structure.box[3:6])
     # Divide by conversion factor
-    box.lengths /= sigma_conversion_factor
```
Doesn't look like it was documented anywhere. Just to have it on record, it looks like this PR also fixes an issue with the lammps writer where the box dimensions were doubly divided by the conversion factor.

daico007 replied:
Yeah, I think before, doing `box.lengths /= sigma_conversion_factor` does not automatically update `box.maxs`, so we have to do it twice.

uppittu11 replied:
Ah. Ok that makes sense.
